python/cloudtik/providers/_private/aliyun/config.py

```diff
@@ -566,28 +568,34 @@ def post_prepare_aliyun(config: Dict[str, Any]) -> Dict[str, Any]:
     return config
 
 
-def list_ecs_instances(provider_config) -> List[Dict[str, Any]]:
+def list_ecs_instances(provider_config) -> Dict[str, Any]:
     """Get all instance-types/resources available.
     Args:
         provider_config: the provider config of the Alibaba Cloud.
     Returns:
-        final_instance_types: a list of instances.
+        final_instance_types: a dict of instance types.
 
     """
-    final_instance_types = []
+    final_instance_types = {}
     ecs_client = EcsClient(provider_config)
     instance_types_body = ecs_client.describe_instance_types()
     if (instance_types_body.instance_types is not None
             and instance_types_body.instance_types.instance_type is not None):
-        final_instance_types.extend(
-            copy.deepcopy(instance_types_body.instance_types.instance_type))
+        for instance_type in instance_types_body.instance_types.instance_type:
+            final_instance_types[instance_type.instance_type_id] = instance_type
+
     while instance_types_body.next_token is not None:
+        prefre_len = len(final_instance_types)
         instance_types_body = ecs_client.describe_instance_types(
             next_token=instance_types_body.next_token)
         if (instance_types_body.instance_types is not None
                 and instance_types_body.instance_types.instance_type is not None):
-            final_instance_types.extend(
-                copy.deepcopy(instance_types_body.instance_types.instance_type))
+            for instance_type in instance_types_body.instance_types.instance_type:
+                final_instance_types[instance_type.instance_type_id] = instance_type
+
+        if len(final_instance_types) == prefre_len:
+            break
```

Comment:
What's the purpose of prefre_len?

Reply:
@jerrychenhf The condition about "instance_types_body.next_token is not None" is always true when the request executed successfully. So it can be found that this function cannot jump out of the loop and will continue to execute.

Reply:
@jerrychenhf I've simplified this code. After reaching the latest page, the value of "instance_types_body.next_token" will be equal to "".